barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp

```diff
@@ -89,18 +89,106 @@ poly_triple serialize_arithmetic_gate(Program::Expression const& arg)
+                   .b = 0,
+                   .c = 0,
+                   .d = 0,
+                   .mul_scaling = 0,
+                   .a_scaling = 0,
+                   .b_scaling = 0,
+                   .c_scaling = 0,
+                   .d_scaling = 0,
+                   .const_scaling = 0 };
+
+    // Flags indicating whether each witness index for the present poly_tuple has been set
+    bool a_set = false;
+    bool b_set = false;
+    bool c_set = false;
+    bool d_set = false;
+    ASSERT(arg.mul_terms.size() <= 1); // We can only accommodate 1 quadratic term
+    // Note: mul_terms are tuples of the form {selector_value, witness_idx_1, witness_idx_2}
+    if (!arg.mul_terms.empty()) {
+        const auto& mul_term = arg.mul_terms[0];
+        quad.mul_scaling = uint256_t(std::get<0>(mul_term));
+        quad.a = std::get<1>(mul_term).value;
+        quad.b = std::get<2>(mul_term).value;
+        a_set = true;
+        b_set = true;
+    }
+    // If necessary, set values for linears terms q_l * w_l, q_r * w_r and q_o * w_o
+    ASSERT(arg.linear_combinations.size() <= 4); // We can only accommodate 4 linear terms
+    for (const auto& linear_term : arg.linear_combinations) {
+        bb::fr selector_value(uint256_t(std::get<0>(linear_term)));
+        uint32_t witness_idx = std::get<1>(linear_term).value;
+
+        // If the witness index has not yet been set or if the corresponding linear term is active, set the witness
+        // index and the corresponding selector value.
+        // TODO(https://github.com/AztecProtocol/barretenberg/issues/816): May need to adjust the pt.a == witness_idx
+        // check (and the others like it) since we initialize a,b,c with 0 but 0 is a valid witness index once the
+        // +1 offset is removed from noir.
+        if (!a_set || quad.a == witness_idx) {
+            quad.a = witness_idx;
+            quad.a_scaling = selector_value;
+            a_set = true;
+        } else if (!b_set || quad.b == witness_idx) {
+            quad.b = witness_idx;
+            quad.b_scaling = selector_value;
+            b_set = true;
+        } else if (!c_set || quad.c == witness_idx) {
+            quad.c = witness_idx;
+            quad.c_scaling = selector_value;
+            c_set = true;
+        } else if (!d_set || quad.d == witness_idx) {
+            quad.d = witness_idx;
+            quad.d_scaling = selector_value;
+            d_set = true;
+        } else {
+            throw_or_abort("Cannot assign linear term to a constraint of width 4");
+        }
+    }
+
+    // Set constant value q_c
+    quad.const_scaling = uint256_t(arg.q_c);
+    return quad;
+}
 
 void handle_arithmetic(Program::Opcode::AssertZero const& arg, AcirFormat& af)
 {
-    af.constraints.push_back(serialize_arithmetic_gate(arg.value));
+    if (arg.value.linear_combinations.size() <= 3) {
+        poly_triple pt = serialize_arithmetic_gate(arg.value);
+        if (pt == poly_triple{ 0, 0, 0, 0, 0, 0, 0, 0 }) {
```

Comment:
I see this logic is why we removed the throw_or_abort in `serialize_arithmetic_gate`, but could we add some comments explaining this logic

Reply:
Good point, I added a comment